packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/s3.ts

```diff
@@ -0,0 +1,39 @@
+import * as iam from '@aws-cdk/aws-iam';
+import * as firehose from '@aws-cdk/aws-kinesisfirehose';
+import { CfnDeliveryStream } from '@aws-cdk/aws-kinesisfirehose';
+import * as s3 from '@aws-cdk/aws-s3';
+import { Construct } from 'constructs';
+
+/**
+ * Props for defining an S3 destination of a Kinesis Data Firehose delivery stream.
+ */
+export interface S3Props extends firehose.DestinationProps { }
+
+/**
+ * An S3 bucket destination for data from a Kinesis Data Firehose delivery stream.
+ */
+export class S3 extends firehose.DestinationBase {
+  constructor(private readonly bucket: s3.IBucket, s3Props: S3Props = {}) {
+    super(s3Props);
+  }
+
+  bind(scope: Construct, options: firehose.DestinationBindOptions): firehose.DestinationConfig {
+    return {
+      properties: {
+        extendedS3DestinationConfiguration: this.createExtendedS3DestinationConfiguration(scope, options.deliveryStream),
+      },
+    };
+  }
+
+  private createExtendedS3DestinationConfiguration(
+    scope: Construct,
+    deliveryStream: firehose.IDeliveryStream,
+  ): CfnDeliveryStream.ExtendedS3DestinationConfigurationProperty {
+    this.bucket.grantReadWrite(deliveryStream);
+    return {
+      cloudWatchLoggingOptions: this.createLoggingOptions(scope, deliveryStream, 'S3Destination'),
+      roleArn: (deliveryStream.grantPrincipal as iam.IRole).roleArn,
+      bucketArn: this.bucket.bucketArn,
+    };
+  }
```

Comment:
The reason this is here is so that my IDE can help me with what properties are available in the ExtendedS3DestinationConfigurationProperty, since that is the DestinationConfig type is just an object. I agree with you that your suggestion is more readable.  

Is there any way to achieve both here? Or do we need to pick one?

Reply:
You can use a local variable inside `bind()` with an explicit type:

```ts
  bind(scope: Construct, options: firehose.DestinationBindOptions): firehose.DestinationConfig {
    this.bucket.grantReadWrite(deliveryStream);
    const s3ExtendendConfig: ExtendedS3DestinationConfigurationProperty = {
      cloudWatchLoggingOptions: this.createLoggingOptions(scope, deliveryStream, 'S3Destination'),
      roleArn: (deliveryStream.grantPrincipal as iam.IRole).roleArn,
      bucketArn: this.bucket.bucketArn,
    };
    return {
      properties: {
        extendedS3DestinationConfiguration: s3ExtendendConfig,
    };
  }
```

Reply:
But I feel like this suggests a bigger issue in the design, that we should perhaps address.

Why doesn't `firehose.DestinationConfig` use strongly-typed L1 interfaces?

Reply:
> I actually agree with this decision. We generally don't want to surface L1 types at the L2 level, and I think places where we do it are a mistake.

I strongly disagree with this statement, and this violates our current L2 guidelines 🙂.

We don't use L1 properties for _input_ properties. But this isn't an input property.

Reply:
Thank you! This is helpful info for me to know the different options here :)

I updated the PR to use a local variable to help with the types. 

@skinny85 To answer:

> But I feel like this suggests a bigger issue in the design, that we should perhaps address.
> 
> Why doesn't `firehose.DestinationConfig` use strongly-typed L1 interfaces?

We did it this way so that the Destination properties can just be injected right into the creation of the CfnDeliveryStream without have to check what type is returned from the bind.

Reply:
Let me know if it would be improved to use the L1 types. There might be a simpler way that I am not thinking of. To me it seems like we would need to do:

```
interface DestinationConfig {
  s3Bucket?: CfnDeliveryStream.ExtendedS3DestinationConfigurationProperty;
  redshiftTable?: CfnDeliverStream.RedshiftDestinationConfigurationProperty;
 ...
}

public bind(...): DestinationConfig {
   return {
       s3Bucket: {....}
  };
}

// then in DeliveryStream constructor:

const destinationConfig = props.destinations[0].bind();

new CfnDeliveryStream(this, 'Resource', {
  deliveryStreamName: props.deliveryStreamName,
  deliveryStreamType: 'DirectPut'.
  extendedS3DestinationConfiguration: destinationConfig.s3Bucket,
  redshiftDestinationConfiguration: destinationConfig.redshiftTable,
  // etc...

Reply:
Yep, that's exactly how I envisioned this. You can even simplify it if you make the names in `DestinationConfig` the same as the names in `CfnDeliveryStreamProps`:

```ts
new CfnDeliveryStream(this, 'Resource', {
  deliveryStreamName: props.deliveryStreamName,
  deliveryStreamType: 'DirectPut',
  ...destinationConfig,
});
```